Introduce shortcut for common line-breaking combinations

fantasai opened the pull request:
#### e59f7bd6c41b92ddd0fce72a1282bb357dbd07d3
<pre>
Introduce shortcut for common line-breaking combinations
<a href="https://bugs.webkit.org/show_bug.cgi?id=273793">https://bugs.webkit.org/show_bug.cgi?id=273793</a>
<a href="https://rdar.apple.com/126007812">rdar://126007812</a>

Reviewed by Alan Baradlay.

Introduces a shortcut to catch the most common character breaking pairs,
avoiding the call to ICU.

Skips any characters that are likely to be tailored.

* LayoutTests/TestExpectations: See bug 274628.
* LayoutTests/platform/ios/TestExpectations: See bug 274628.
* LayoutTests/platform/mac/TestExpectations: See bug 274628.
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::findNextBreakablePosition): Pass in auto-phrase.
* Source/WebCore/rendering/BreakLines.h:
(WebCore::BreakLines::CharacterInfo::CharacterInfo): Add helper datatype.
(WebCore::BreakLines::CharacterInfo::set): Add helper datatype.
(WebCore::BreakLines::CharacterInfo::operator CharacterType const): Add helper datatype.
(WebCore::BreakLines::nextBreakablePosition): Add fast path.
(WebCore::BreakLines::classify): Add UAX14 classification function.
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::computePreferredLogicalWidths): Check for auto-phrase.
* Source/WebCore/rendering/line/BreakingContext.h:
(WebCore::BreakingContext::handleText): Check for auto-phrase.

Canonical link: <a href="https://commits.webkit.org/279318@main">https://commits.webkit.org/279318@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0d6c2179cd9464f4c04dd05e2c8ed090a05d06db

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53104 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32441 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5591 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56383 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3827 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55409 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39314 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3580 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43057 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2474 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55202 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30196 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45835 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24186 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27199 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3164 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/1986 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49050 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3322 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/57979 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28245 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3290 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50459 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46062 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49764 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11590 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30384 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29219 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -77,6 +77,7 @@ ipc/webpageproxy-correctionpanel-no-crash.html [ Skip ]
 # iOS 17 and above support word-break: auto-phrase.
 # Individually failing tests should be marked in separate expectations.
 imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase [ Pass ]
+imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase/word-break-auto-phrase-009.html [ ImageOnlyFailure ]
```
this looks redundant to me.

fantasai replied:
This one isn't redundant, the master file one is actually; but it's in the wrong place. Moved.

alanbaradlay commented on LayoutTests/platform/mac/TestExpectations:
```diff
@@ -104,6 +104,8 @@ media/media-source/remoteplayback-from-source-element.html [ Pass ]
 # macOS 14 and above support word-break: auto-phrase.
 # Individually failing tests should be marked in separate expectations.
 [ Sonoma+ ] imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase [ Pass ]
+imported/w3c/web-platform-tests/css/css-text/word-break/auto-phrase/word-break-auto-phrase-009.html [ ImageOnlyFailure ]
```
same

fantasai replied:
This one isn't redundant, the master file one is actually; but it's in the wrong place. Moved.

alanbaradlay commented on Source/WebCore/rendering/BreakLines.h:
```diff
@@ -39,33 +39,81 @@ class BreakLines {
+    }
+
+    // Data types.
+    enum BreakClass : int {
```
uint16_t?

fantasai replied:
Done.

alanbaradlay commented on Source/WebCore/rendering/BreakLines.h:
```diff
@@ -39,33 +39,81 @@ class BreakLines {
+    template<typename CharacterType>
+    struct CharacterInfo {
+        CharacterType id;
+        BreakClass type;
```
maybe BreakClass type { kIndeterminate };
CharacterType id { 0 };

fantasai replied:
Sure.

fantasai committed:
Introduce shortcut for common line-breaking combinations
https://bugs.webkit.org/show_bug.cgi?id=273793
rdar://126007812

Reviewed by Alan Baradlay.

Introduces a shortcut to catch the most common character breaking pairs,
avoiding the call to ICU.

Skips any characters that are likely to be tailored.

* LayoutTests/TestExpectations: See bug 274628.
* LayoutTests/platform/ios/TestExpectations: See bug 274628.
* LayoutTests/platform/mac/TestExpectations: See bug 274628.
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::findNextBreakablePosition): Pass in auto-phrase.
* Source/WebCore/rendering/BreakLines.h:
(WebCore::BreakLines::CharacterInfo::CharacterInfo): Add helper datatype.
(WebCore::BreakLines::CharacterInfo::set): Add helper datatype.
(WebCore::BreakLines::CharacterInfo::operator CharacterType const): Add helper datatype.
(WebCore::BreakLines::nextBreakablePosition): Add fast path.
(WebCore::BreakLines::classify): Add UAX14 classification function.
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::computePreferredLogicalWidths): Check for auto-phrase.
* Source/WebCore/rendering/line/BreakingContext.h:
(WebCore::BreakingContext::handleText): Check for auto-phrase.

Canonical link: https://commits.webkit.org/279318@main
